File: structural/typescript/adapter_structured/adapters.ts
import { MediaPlayer, PaymentProcessor, CelsiusTemperature, FahrenheitTemperature } from './interfaces';
import { AdvancedMediaPlayer, StripePaymentSystem, PayPalPaymentSystem } from './adaptees';

export class MediaAdapter implements MediaPlayer {
  private advancedMusicPlayer: AdvancedMediaPlayer;
  
  constructor() { this.advancedMusicPlayer = new AdvancedMediaPlayer(); }
  
  play(audioType: string, fileName: string): void {
    if (audioType === 'vlc') { this.advancedMusicPlayer.playVlc(fileName); }
    else if (audioType === 'mp4') { this.advancedMusicPlayer.playMp4(fileName); }
    else { console.log(`Invalid media. ${audioType} format not supported`); }
  }
}

export class StripeAdapter implements PaymentProcessor {
  private stripeSystem: StripePaymentSystem;
  constructor() { this.stripeSystem = new StripePaymentSystem(); }
  processPayment(amount: number): void { this.stripeSystem.makeStripePayment(amount); }
}

export class PayPalAdapter implements PaymentProcessor {
  private payPalSystem: PayPalPaymentSystem;
  constructor() { this.payPalSystem = new PayPalPaymentSystem(); }
  processPayment(amount: number): void { this.payPalSystem.sendPayPalRequest(amount); }
}

export class TemperatureAdapter implements CelsiusTemperature, FahrenheitTemperature {
  private temperatureInCelsius: number;
  constructor(temp: number = 0) { this.temperatureInCelsius = temp; }
  getTemperatureInCelsius(): number { return this.temperatureInCelsius; }
  setTemperatureInCelsius(temp: number): void { this.temperatureInCelsius = temp; }
  getTemperatureInFahrenheit(): number { return (this.temperatureInCelsius * 9 / 5) + 32; }
  setTemperatureInFahrenheit(temp: number): void { this.temperatureInCelsius = (temp - 32) * 5 / 9; }
}

File: structural/dart/adapter_structured/adapters.dart
import 'interfaces.dart';
import 'adaptees.dart';

class MediaAdapter implements MediaPlayer {
  final AdvancedMediaPlayer _advancedMusicPlayer = AdvancedMediaPlayer();
  
  @override
  void play(String audioType, String fileName) {
    if (audioType == 'vlc') {
      _advancedMusicPlayer.playVlc(fileName);
    } else if (audioType == 'mp4') {
      _advancedMusicPlayer.playMp4(fileName);
    } else {
      print('Invalid media. $audioType format not supported');
    }
  }
}

class StripeAdapter implements PaymentProcessor {
  final StripePaymentSystem _stripeSystem = StripePaymentSystem();
  
  @override
  void processPayment(double amount) {
    _stripeSystem.makeStripePayment(amount);
  }
}

class PayPalAdapter implements PaymentProcessor {
  final PayPalPaymentSystem _payPalSystem = PayPalPaymentSystem();
  
  @override
  void processPayment(double amount) {
    _payPalSystem.sendPayPalRequest(amount);
  }
}

class TemperatureAdapter implements CelsiusTemperature, FahrenheitTemperature {
  double _temperatureInCelsius;
  
  TemperatureAdapter([this._temperatureInCelsius = 0.0]);
  
  @override
  double getTemperatureInCelsius() {
    return _temperatureInCelsius;
  }
  
  @override
  void setTemperatureInCelsius(double temp) {
    _temperatureInCelsius = temp;
  }
  
  @override
  double getTemperatureInFahrenheit() {
    return (_temperatureInCelsius * 9 / 5) + 32;
  }
  
  @override
  void setTemperatureInFahrenheit(double temp) {
    _temperatureInCelsius = (temp - 32) * 5 / 9;
  }
}
